### ISSUE TITLE
Feature Request: Rule to avoid module coupling

### ISSUE THREAD

**Issue by ndrluis** [astral-sh/ruff#10300]
Hello, I'm not a Pythonist, but I'm working on some Python projects and having a problem with how to avoid some import coupling. 

For example:
```python
# module_a.py

def func_a():
    pass

# module_b.py

from module_a import func_a

def func_b():
    func_a()

# module_c.py

from module_b import func_a  # This line should import func_a directly from module_a instead

def func_c():
    func_a()
```    
  
We want to avoid having `func_a` be imported from `module_b`; instead, we intend to import it from the original module where `func_a` is defined. I know that we could use the __all__ variable to explicitly "close the door," but I believe that a rule that warns us when this is occurring would be better.

Any thoughts on this?

**Comment by mkniewallner** [comment]
I believe that the feature would be similar to what https://github.com/seddonym/import-linter provides, right? Although less powerful than the library, in Ruff you should be able to achieve something similar with https://docs.astral.sh/ruff/rules/banned-api/.

https://github.com/astral-sh/ruff/issues/1912 exists to implement `import-linter` in Ruff as well.

**Comment by eli-schwartz** [comment]
I recall searching for this feature and coming across import-linter, then disregarding it since import-linter seems to do something else entirely -- and require significant configuration to do so.

The problem is not actually about allowing or banning specific APIs from being used. The problem is that if you have a file that includes `import os`, "os" is a validly existing symbol reference inside your module and "import mymodule.os" works at runtime...

... but no one would actually consider this a reasonable usage of the "os" module. Because it's an import.

This can be trivially detected in a heuristic manner, you don't need to define rulesets for specific examples you don't want to use. All you need to do is check whether your module's public interface is being obeyed. It's intrinsically tied to `__all__`.

It turns out that this works well together with static typing, so mypy includes an option for it: https://mypy.readthedocs.io/en/stable/config_file.html#confval-implicit_reexport

I enabled the mypy option and caught a looooooot of cases where we were doing it in our codebase. :D

**Comment by eli-schwartz** [comment]
It sounds potentially interesting for ruff to implement that either in addition to or instead of import-linter. Unlike import-linter it has the advantage of by design not requiring any sort of configuration.
